src/main/kotlin/com/openlattice/chronicle/services/notifications/NotificationService.kt

```diff
@@ -0,0 +1,141 @@
+import com.openlattice.chronicle.storage.PostgresColumns.Companion.UPDATED_AT
+import com.openlattice.chronicle.storage.StorageResolver
+import org.slf4j.LoggerFactory
+import org.springframework.stereotype.Service
+import java.sql.Connection
+import java.sql.SQLException
+import java.time.OffsetDateTime
+import java.util.concurrent.ExecutionException
+
+
+/**
+ * @author Todd Bergman <todd@openlattice.com>
+ */
+
+@Service
+class NotificationService(
+    private val storageResolver: StorageResolver,
+    private val authorizationService: AuthorizationManager,
+    private val twilioService :TwilioService,
+    override val auditingManager: AuditingManager,
+) : NotificationManager, AuditingComponent {
+
+    companion object {
+        private val logger = LoggerFactory.getLogger(NotificationService::class.java)
+
+        private val NOTIFICAITON_COLUMNS = NOTIFICATIONS.columns.joinToString(",") { it.name }
+
+        private val UPDATE_NOTIFICATION_COLUMNS = listOf(
+            UPDATED_AT,
+            STATUS,
+        ).joinToString(",") { it.name }
+
+        private val INSERT_NOTIFICATION_SQL = """
+            INSERT INTO ${NOTIFICATIONS.name} (${NOTIFICAITON_COLUMNS}) VALUES (?,?,?,?,?,?,?,?,?,?,?,?)
+        """.trimIndent()
+
+        private val GET_NOTIFICATION_ID_FROM_MESSAGE_ID_SQL = "SELECT * FROM ${NOTIFICATIONS.name} WHERE ${MESSAGE_ID.name} = ?"
+
+        private val UPDATE_NOTIFICATION_SQL = """
+            UPDATE ${NOTIFICATIONS.name}
+            SET (${UPDATE_NOTIFICATION_COLUMNS}) = (?,?)
+            WHERE ${NOTIFICATION_ID.name} = ?
+        """.trimIndent()
+    }
+
+    private fun insertNotifications(connection: Connection, notifications: List<Notification>): Int {
+        val ps = connection.prepareStatement(INSERT_NOTIFICATION_SQL)
+        try {
+            notifications.forEach { notification ->
+                ps.setObject(1, notification.id)
+                ps.setObject(2, notification.candidateId)
+                ps.setObject(3, notification.organizationId)
+                ps.setObject(4, notification.studyId)
+                ps.setObject(5, notification.createdAt)
+                ps.setObject(6, notification.updatedAt)
+                ps.setObject(7, notification.messageId)
+                ps.setObject(8, notification.type)
+                ps.setObject(9, notification.status)
+                ps.setObject(10, notification.body)
+                ps.setObject(11, notification.email)
+                ps.setObject(12, notification.phone)
+                ps.addBatch()
+                authorizationService.createSecurableObject(
+                    connection = connection,
+                    aclKey = AclKey(notification.id),
+                    principal = Principals.getCurrentUser(),
+                    objectType = SecurableObjectType.Notification
+                )
+            }
+            return ps.executeBatch().sum()
+        } catch (e :SQLException) {
+            throw Error(e)
```

Comment:
Don't throw Error. (a) it is not an unrecoverable occurrence and (b) we don't to terminate JVM.

Reply:
The correct pattern here is to do connection.prepareStatement(INSERT_NOTIFICATION_SQL).use { ... } which will close the statement if there's an expcetion. You can also move the authorization.createSecurableObject outside of the use clause as it will be outside of connection. If you want to be really safe add a check( !connection.autoCommit ) {"Auto-commit must be disabled on provided connection"}. This can even be implemented in a Util class so that other similar code just call ensureNoAutoCommit(connection).

Reply:
i.e let SQLException bubble up or if you must wrap with something that is handled by Chronicle error handler.

Reply:
was using try catch to debug - I'm just gonna remove it.